Content.Server/Radio/EntitySystems/HeadsetSystem.cs

```diff
@@ -78,29 +100,116 @@ public void SetEnabled(EntityUid uid, bool value, HeadsetComponent? component =
 
     private void OnHeadsetReceive(EntityUid uid, HeadsetComponent component, RadioReceiveEvent args)
     {
-        if (TryComp(Transform(uid).ParentUid, out ActorComponent? actor))
+        if(TryComp(Transform(uid).ParentUid, out ActorComponent? actor))
             _netMan.ServerSendMessage(args.ChatMsg, actor.PlayerSession.ConnectedClient);
     }
 
     private void OnExamined(EntityUid uid, HeadsetComponent component, ExaminedEvent args)
     {
-        if (!args.IsInDetailsRange)
+        if(!args.IsInDetailsRange)
             return;
+        if(component.KeysInstalledAmount == 0)
+        {
+            args.PushMarkup(Loc.GetString("examine-headset-no-keys"));
+        }
+        if(component.Channels.Count > 0)
+        {
+            args.PushMarkup(Loc.GetString("examine-headset"));
+            foreach(var id in component.Channels)
+            {
+                // if(id == "Common")
+                //     continue;
+                var proto = _protoManager.Index<RadioChannelPrototype>(id);
+                args.PushMarkup(Loc.GetString("examine-headset-channel",
+                    ("color", proto.Color),
+                    ("key", proto.KeyCode),
+                    ("id", proto.LocalizedName),
+                    ("freq", proto.Frequency)));
+            }
+            args.PushMarkup(Loc.GetString("examine-headset-chat-prefix", ("prefix", ";")));
+        }
+    }
+
+    private void OnStartup(EntityUid uid, HeadsetComponent component, ComponentStartup args)
+    {
+        component.KeyContainer = _container.EnsureContainer<Container>(uid, HeadsetComponent.KeyContainerName);
+    }
+
+    private bool InstallKey(HeadsetComponent src, EntityUid key)
+    {
+        if(src.KeyContainer.Insert(key))
+        {
+            ++src.KeysInstalledAmount;
+            return true;
+        }
+        return false;
+    }
 
-        args.PushMarkup(Loc.GetString("examine-headset"));
+    private void RecalculateChannels(HeadsetComponent src)
+    {
+        src.Channels.Clear();
+        src.KeysInstalledAmount = 0;
+        foreach(EntityUid i in src.KeyContainer.ContainedEntities)
+            if(TryComp<EncryptionKeyComponent?>(i, out var key))
+            {
+                ++src.KeysInstalledAmount;
+                foreach(var j in key.Channels)
+                    src.Channels.Add(j);
+            }
+        return;
```

Comment:
As above, also please just use `var`, don't explicitly type for stuff like this, and use an entityquery outside of the loop.

Reply:
What entityquery